library-api/src/main/java/au/csiro/pathling/library/data/ReadableSource.java

```diff
@@ -0,0 +1,176 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package au.csiro.pathling.library.data;
+
+import static java.util.Objects.requireNonNull;
+
+import au.csiro.pathling.aggregate.AggregateQueryExecutor;
+import au.csiro.pathling.aggregate.AggregateRequest;
+import au.csiro.pathling.config.QueryConfiguration;
+import au.csiro.pathling.extract.ExtractQueryExecutor;
+import au.csiro.pathling.extract.ExtractRequest;
+import au.csiro.pathling.library.PathlingContext;
+import au.csiro.pathling.library.query.AggregateQuery;
+import au.csiro.pathling.library.query.ExtractQuery;
+import au.csiro.pathling.query.DataSource;
+import au.csiro.pathling.terminology.TerminologyServiceFactory;
+import ca.uhn.fhir.context.FhirContext;
+import java.util.Optional;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import lombok.Value;
+import lombok.experimental.NonFinal;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.hl7.fhir.r4.model.Enumerations.ResourceType;
+
+/**
+ * The class representing a client that can execute complex queries such as extract or aggregate on
+ * a {@link DataSource}.
+ */
+@Value
+@NonFinal
+public class ReadableSource {
+
+  @Nonnull
+  FhirContext fhirContext;
+
+  @Nonnull
+  SparkSession sparkSession;
+
+  @Nonnull
+  DataSource dataSource;
+
+  @Nonnull
+  QueryConfiguration configuration;
+
+  @Nonnull
+  Optional<TerminologyServiceFactory> terminologyClientFactory;
+
+  /**
+   * A builder of {@link ReadableSource} that allows for explicit assignment of the underlying
+   * {@link DataSource}.
+   */
+  public static class Builder extends AbstractClientBuilder<Builder> {
+
+    @Nullable
+    private DataSource dataSource;
+
+    protected Builder(@Nonnull final PathlingContext pathlingContext) {
+      super(pathlingContext);
+    }
+
+    @Nonnull
+    @Override
+    protected DataSource buildDataSource() {
+      return requireNonNull(dataSource);
+    }
+
+    /**
+     * Sets the {@link DataSource} the client should be using.
+     *
+     * @param dataSource the data source.
+     * @return this builder.
+     */
+    @Nonnull
+    public Builder withDataSource(@Nonnull final DataSource dataSource) {
+      this.dataSource = dataSource;
+      return this;
+    }
+  }
+
+  /**
+   * Creates a new extract query bound to this client with given subject resource type.
+   *
+   * @param subjectResourceType the type of subject resource for the query.
+   * @return the new instance of extract query.
+   */
+  @Nonnull
+  public ExtractQuery newExtractQuery(@Nonnull final ResourceType subjectResourceType) {
```

Comment:
I think the examples would read more naturally if these methods were named more succinctly:

- `newExtractQuery` -> `extract`
- `newAggregateQuery` -> `aggregate`

Then the example would look more like this:

```java
    final Dataset<Row> patientResult = readableSource.aggregate(ResourceType.PATIENT)
        .withGrouping("gender")
        .withGrouping("maritalStatus.coding.code", "maritalStatusCode")
        .withAggregation("count()", "patientCount")
        .withAggregation("id.count()")
        .withFilter("birthDate > @1957-06-06")
        .execute();
```

Reply:
ok